source/extensions/filters/http/fault/fault_filter.cc

```diff
@@ -318,8 +357,101 @@ void FaultFilter::resetTimerState() {
   }
 }
 
-void FaultFilter::setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) {
-  callbacks_ = &callbacks;
+Http::FilterDataStatus FaultFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (response_limiter_ != nullptr) {
+    response_limiter_->writeData(data, end_stream);
+    return Http::FilterDataStatus::StopIterationNoBuffer;
+  }
+
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus FaultFilter::encodeTrailers(Http::HeaderMap&) {
+  if (response_limiter_ != nullptr) {
+    return response_limiter_->onTrailers();
+  }
+
+  return Http::FilterTrailersStatus::Continue;
+}
+
+StreamRateLimiter::StreamRateLimiter(uint64_t max_kbps, uint64_t max_buffered_data,
+                                     std::function<void()> pause_data_cb,
+                                     std::function<void()> resume_data_cb,
+                                     std::function<void(Buffer::Instance&, bool)> write_data_cb,
+                                     std::function<void()> continue_cb, TimeSource& time_source,
+                                     Event::Dispatcher& dispatcher)
+    : // bytes_per_time_slice is KiB converted to bytes divided by the number of ticks per second.
+      bytes_per_time_slice_((max_kbps * 1024) / SecondDivisor), write_data_cb_(write_data_cb),
+      continue_cb_(continue_cb),
+      // The token bucket is configured with a max token count of the number of ticks per second,
+      // and refills at the same rate, so that we have a per second limit which refills gradually in
+      // ~63ms intervals.
```

Comment:
I think I just remembered something about the backing issue for this PR.. this is bandwidth rate limiting in essence.. but at application layer. No guarantees that the OS isn't buffering it up until MSS and then pushing data [granted there is a timer in the OS, that will cap the wait time]. I used to do this with `netem` module in the kernel which has a bewildering array of such algorithms, but applied at the netdev level, after TCP packaging. It would be interesting to see if the OS will mask the user space bandwidth throttling or not..

Reply:
also how did you come up with the 63ms figure?

Reply:
re: your first comment, if there is a concern can you restate? 

re: 63ms, 1000ms / 16 (see other comments).